src/vector_of_array.jl

```diff
@@ -0,0 +1,65 @@
+abstract AbstractVectorOfArray{T, N} <: AbstractArray{T, N}
+
+# Based on code from M. Bauman Stackexchange answer + Gitter discussion
+type VectorOfArray{T, N, A} <: AbstractVectorOfArray{T, N}
+  data::A # A <: AbstractVector{<: AbstractArray{T, N - 1}}
+  dims::NTuple{N, Int}
+  ragged::Bool
+end
+
+function VectorOfArray(vec::AbstractVector)
+  # Only allow a vector of equally shaped subtypes, this will not work for ragged arrays
+  ragged = !all(size(vec[1]) == size(v) for v in vec)
+  VectorOfArray(vec, (size(vec[1])..., length(vec)), ragged)
+end
+
+VectorOfArray{T, N}(vec::AbstractVector{T}, dims::NTuple{N}, ragged::Bool) = VectorOfArray{eltype(T), N, typeof(vec)}(vec, dims, ragged)
+@inline function Base.size(S::AbstractVectorOfArray)
+  if S.ragged
+    return size(S.data)
+  else
+    return S.dims
+  end
+end
+
+@inline function Base.getindex{T, N}(S::AbstractVectorOfArray{T, N}, I::Vararg{Int, N})
+  @boundscheck checkbounds(S, I...)
+  S.ragged && throw(BoundsError("A ragged VectorOfArray does not support Cartesian indexing"))
+  S.data[I[end]][Base.front(I)...]
+end
+# Linear indexing will be over the container elements, not the individual elements
+# unlike an true AbstractArray
+@inline Base.getindex{T, N}(S::AbstractVectorOfArray{T, N}, I::Union{Int, Colon, AbstractArray{Int}}) = S.data[I]
+
+Base.copy(S::AbstractVectorOfArray) = VectorOfArray(copy(S.data), S.dims, S.ragged)
+
+Base.sizehint!{T, N}(S::AbstractVectorOfArray{T, N}, i) = sizehint!(S.data, i)
+
+function Base.push!{T, N}(S::AbstractVectorOfArray{T, N}, new_item::AbstractArray)
+  if S.dims[1:(end - 1)] == size(new_item)
```

Comment:
We are updating the dims attribute if it matches the current shape of the other elements (we drop the last element as this is the length of the containing vector). Otherwise we make it ragged. But you show a bug, I need to check that it has already become ragged in which case I don't need to do anymore checking. Will fix this up.